00_Base/src/util/middleware/ResponseValidationMiddleware.ts

```diff
@@ -0,0 +1,90 @@
+import {
+  getMetadataArgsStorage,
+  InternalServerError,
+  KoaMiddlewareInterface,
+  Middleware,
+} from 'routing-controllers';
+import { plainToClass } from '../util';
+import { Service } from 'typedi';
+import { Context } from 'vm';
+import { validate } from 'class-validator';
+import { validatedResponseParam } from '../decorators/ValidatedResponseSchema';
+import {
+  OcpiResponse,
+  OcpiResponseStatusCode,
+  OcpiRoutePrefix,
+} from '../../index';
+import { ResponseValidationError } from '../../exception/ResponseValidationError';
+
+@Middleware({ type: 'before' })
+@Service()
+export class ResponseValidationMiddleware implements KoaMiddlewareInterface {
+  async use(
+    context: Context,
+    next: (err?: any) => Promise<any>,
+    ...args: any[]
+  ): Promise<any> {
+    await next();
+
+    const route = this.findRoute(context);
+    if (route) {
+      const { target, method } = route;
+      const responseClass = Reflect.getMetadata(
+        validatedResponseParam,
+        target.prototype,
+        method,
+      );
+
+      if (responseClass && context.body) {
+        const dto: OcpiResponse<any> = plainToClass(
+          responseClass,
+          context.body,
+        );
+        if (dto.status_code === OcpiResponseStatusCode.GenericSuccessCode) {
+          const errors = await validate(dto);
+          if (errors.length > 0) {
+            const errorMessages = errors
+              .map((error) => error.toString())
+              .join('');
+            throw new ResponseValidationError(errorMessages);
+          }
+        } else {
+          // todo handle OcpiErrorResponse and OcpiEmptyResponse
+        }
+      }
+    } else {
+      throw new InternalServerError(
+        'Could not find matching route? This should not happen',
+      );
+    }
+  }
+
+  private findRoute = (ctx: Context) => {
+    const storage = getMetadataArgsStorage();
+    const matchedRoute = ctx._matchedRoute;
+    if (!matchedRoute) return null;
+    const method = ctx.request?.req?.method?.toLowerCase();
+    if (!method) return null;
+    const controllerMetadata = storage.controllers.find((controller) => {
+      return matchedRoute.startsWith(`${OcpiRoutePrefix}${controller.route}`);
```

Comment:
How does this work for paths with params?

Reply:
Good question - so it is best to loo at what `getMetadataArgsStorage()` provides, but essentially in there, among other things, it has a list of `controllers` that look like this:
```
{
  "type": "json",
  "target: Constructible,
  "route": "/:versionId/locations"
}
```
and `actions` that looks like this: 
```
{
  "type": "get",
  "method": "getLocationById",
  "route": "/:location_id"
  "target: Constructible,
},
{
  "type": "get",
  "method": "getLocations"
  "target: Constructible,
}
```
you can see that actions may or may not have a path. The code in this PR gets a `matchingRoute` from the context and is then able to match the route in the controller plus the action to figure out which method is actually being called. We need to do this so that we can call :
```
const responseClass = Reflect.getMetadata(
        validatedResponseParam,
        target.prototype,
        method,
      );
```
to get the actual class that was passed into the `@ValidatedResponseSchema` annotation which would look like this:
```
@ValidatedResponseSchema(LocationResponse, {
    statusCode: HttpStatus.OK,
    description: 'Successful response',
    examples: {
      success: MOCK_LOCATION,
    },
  })
```